### ISSUE TITLE
Managers: driver methods return type not used

### ISSUE THREAD

**Issue by jf-m** [nunomaduro/larastan#2046]
- Larastan Version: 2.9.8
- Laravel Version: 10

### Description
When dealing with Managers, the class of the object returned by the `->driver()` method is used instead of the return type defined in the `createFooDriver()` method.

If a contract is defined in the return type of the `createFooDriver`, larastan will not consider that the manager returns the contract, but the actual implementation itself instead.

Example:

If we have the following Manager, Contract and driver:

```php
class Storage extends Manager
{
    public function getDefaultDriver(): string {
      return 'file';
    }

    public function createFileDriver(): StorageContract {
      return app(FileDriver::class);
    }
}

interface StorageContract
{
    public function type(): string;
}

class FileDriver implements StorageContract
{
    public function type(): string {
      return 'file';
    }

    public function helloWorld(): string {
       return 'Hello World';
     }
}
```
Then I expect the following code to fail with larastan:

```php
app(Storage::class)->helloWorld();
```

Instead, this code pass the checks.

## Solution
I believe the following line of code can be improved : https://github.com/larastan/larastan/blob/2.x/src/Methods/Pipes/Managers.php#L39

Instead of simply using the driver's class, we should get the return type of the Manager's `'create'.Str::studly($concrete->getDefaultDriver()).'Driver';` method.

I can submit a PR with a fix.

Thank you very much for maintaining this awesome project :+1:

**Comment by calebdw** [comment]
This is intentional. Like it or not, Larastan resolves the implementation for you instead of the contract for most things in the framework.

**Comment by jf-m** [comment]
Alright, I understand, thanks for your reply :+1:

**Comment by jf-m** [comment]
@calebdw 
Do you think this could still be implemented with a custom config parameter? 

Something like `resolveImplementations: false` would then resolve contracts instead of implementations. 
This configuration would have an impact not only on the manager but on all the places where larastan resolves the implementations.

**Comment by calebdw** [comment]
It could possibly be implemented, but at the cost of additional complexity. I'm not sure if it is worth the additional complexity...

The Laravel community has a bad habit of typing contracts and then calling methods that don't exist on the contract but the runtime implementation. Part of the problem I think is that adding methods to the contract is seen as a BC, so the methods are added to the runtime implementation (because people want to use them now) and then they forget to add to the contract in the next major release. In particular, the `\Illuminate\Contracts\Foundation\Application` and `\Illuminate\Foundation\Application` are bad about this.

For Manager/Drivers like this, I'm not sure I really see a need for a Driver to contain public methods that aren't on the contract (since you would never typehint a specific driver in your code)? Granted, nothing in Laravel forces your drivers to implement a contract in the first place.

Something I have been thinking about though is creating a way to enforce that a class should not contain any public methods that don't exist on the contract(s)

**Comment by jf-m** [comment]
I agree, there's no point having a public method on a driver that is not on the contract.

Creating a way to enforce this as a rule in larastan would solve this issue and would work for this driver/manager case, but for other cases and with people who have this bad habit of calling methods that don't exist on the contract,... they probably won't like it :grimacing: . 
I mean, you said yourself that in some situations, they do it to avoid BC regardless of it being a bad habit. But I don't think anyone would do that in a Driver/Manager situation because it doesn't make sense as the implementation is never typehinted like you said.

In my opinion, the Driver/Manager use case is definitely a case where the contract (if there's one) **must** be enforced. Like you said, there's absolutely no reason one would add a public method to the runtime implementation without adding it to the contract because of the nature of this Driver/Manager feature. This would be a weakness in the code that I think larastan could/should catch and warn about.


Do you know on the top of your head what are the other cases in larastan where it's been decided to consider the runtime implementation over the contract? Maybe we'll find out that the Driver/Manager case is the only one of this "kind"?

**Comment by calebdw** [comment]
> Something I have been thinking about though is creating a way to enforce that a class should not contain any public methods that don't exist on the contract(s)

This would be opt-in on a per-class basis. I'm thinking something like a `#[EnforceContractMethods]` PHP attribute that can mark specific classes.

However, this is more of an architecture test so it might be something that @carlosas could implement in [phpat](https://github.com/carlosas/phpat)---something like:

```php
    #[Test]
    public function storage_drivers_have_no_extra_public_methods(): Rule
    {
        return PHPat::rule()
            ->classes(Selector::namespace('App\Storage\Drivers'))
            ->shouldHaveAllPublicMethodsCoveredByContracts()
            ->classes(Selector::classname(StorageContract::class));
    }
```

**Comment by calebdw** [comment]
@jf-m,

After thinking about it, I suppose using the driver creation return type would work---then the end user can decide what they want:

```php
// if you want the contract
public function createFileDriver(): StorageContract;
// or if you want the implementation
public function createFileDriver(): FileDriver;
```

**Comment by jf-m** [comment]
@calebdw 

Yes, that was my initial thought. If the user decides to put a contract as a return type, we should enforce it in larastan.

I can do the PR if that's alright with you?

**Comment by calebdw** [comment]
You can do the PR if you want, or I can take a crack at it. Just make sure to also write tests---if you need help just let me know :)
